crates/block-producer/src/batch_builder/mod.rs

```diff
@@ -206,3 +221,106 @@ where
         Ok(())
     }
 }
+
+pub struct BatchProducer {
+    pub batch_interval: Duration,
+    pub workers: NonZeroUsize,
+    pub mempool: Arc<Mutex<Mempool>>,
+    pub tx_per_batch: usize,
+    pub simulated_proof_time: Duration,
+}
+
+type BatchResult = Result<(BatchJobId, TransactionBatch), (BatchJobId, BuildBatchError)>;
+
+/// Wrapper around tokio's JoinSet that remains pending if the set is empty,
+/// instead of returning None.
+struct WorkerPool {
+    in_progress: JoinSet<BatchResult>,
+    simulated_proof_time: Duration,
+}
+
+impl WorkerPool {
+    fn new(simulated_proof_time: Duration) -> Self {
+        Self {
+            simulated_proof_time,
+            in_progress: JoinSet::new(),
+        }
+    }
+
+    async fn join_next(&mut self) -> Result<BatchResult, tokio::task::JoinError> {
+        if self.in_progress.is_empty() {
+            std::future::pending().await
+        } else {
+            // Cannot be None as its not empty.
+            self.in_progress.join_next().await.unwrap()
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.in_progress.len()
+    }
+
+    fn spawn(&mut self, id: BatchJobId, transactions: Vec<ProvenTransaction>) {
+        self.in_progress.spawn({
+            let simulated_proof_time = self.simulated_proof_time;
+            async move {
+                tracing::debug!("Begin proving batch.");
+
+                let batch = TransactionBatch::new(transactions, Default::default())
+                    .map_err(|err| (id, err))?;
+
+                tokio::time::sleep(simulated_proof_time).await;
```

Comment:
Question: is this going to sleep for the exact amount of time? I'm wondering if we should add some variance to it.

Reply:
We can define some distribution function if you like :) I assume just a uniform distribution between a min and a max would be good enough?

Reply:
#519